### PR TITLE
Ownership check with security domains 

### PR DESCRIPTION
**Pull request by albertzaharovits**
EDIT 10.01.2022: The below is not currently an accurate description, see https://github.com/elastic/elasticsearch/pull/82335#issuecomment-1008903642 instead:

-----

This is under construction and will be split into components, but the general plan is:

Owned resources: Profiles, API Keys, Scroll Searches, Async Search Results, Refresh tokens

If the current username accessing some resource IS NOT under any domain (no domain configured for the realm that authenticated/looked-up the user), then this user can access the resources created under all/any domains, but only owned by the same username of the same realm name and type.

If the current username accessing some resource IS NOT under any domain (no domain configured for the realm that authenticated/looked-up the user), then this user can access the resources owned by the same username of the same realm name and type, even if these resources have been created while
the realm was assigned to a domain.

If the current username accessing the resource IS under some domain (the realm that authenticated/looked-up the user is assigned to a domain), then this user can access the resources of the same username created under the same domain (no matter the realms name and type), as well as the resources owned by the same username of the same realm name and type, created when the realm was domain-less.

The above implies that:
 * a user from a domain-less realm has access to its resources as long as the realm is not renamed; if the realm is renamed the user loses access to its resources created before the time of the rename
 * a user continues to have access to its resources after the realm is assigned to any domain(s), as long as the realm is not renamed
 * if the realm is renamed after it's being assigned to a domain, the user continues to have access to the resources created after the realm was assigned to the domain, but loses access to the resources created under the domain-less realm
 * if the domain is renamed, the user loses access to all the resources created while the realm was assigned to the old domain, but continues to have access to the resources created when the realm was domain-less (assuming the realm has not been renamed)
 * if the user is authenticated/looked-up by another realm, with the same username and under the same domain,
 it has access to the resources created under the other realm(s), but only to those created while the realm was under the same domain (excluding those created when the realm was domain-less, or under a different realm)

### CODE REVIEW

**Comment by albertzaharovits** [comment]
The code as is today at https://github.com/elastic/elasticsearch/pull/82335/commits/5bf76bfe21c4b8ac64aadb66a61a82dc039353b3 ,
only represents a big refactoring with the goal to garner all the "ownership" checks inside the `SecurityContext` class.
There's no code yet commited that actually implements the domain-wise ownership check.

**Comment by albertzaharovits** [comment]
@ywang made me realize that the above goes against our plan from last time, that I myself was advancing.
I think I lost track of that, between me diving into the code and looking at API Keys rather than profiles as well as thinking about realm rename.

**TLDR We're going to store ONLY the realm name/type on the entities. When handling an API for a resource, the current realm to domain assignment is all that is relevant in order to determine if the resource is accessible.**

In principle this means that ownership is maintained based on realm name/type, as it is also currently the case.
Domain association only allows for the mutual visibility of resources among the realms in a domain (this is how we choose to define a domain).

Longer story:
At one point in time, a realm can be associated with one domain only.
But it is technically possible to change the domain association during the lifetime of a realm.
In this case the question that arises is what happens to the resources that were created, or to those that were just visible, before the realm changes domain association.
We concluded that it can actually be argued both ways: the resources can remain with the old domain or roam to the new one.
Moreover both imply that there is a separation between entities that are owned and entities that are just visible, although this separation is not explicit when the resources are created (to the extent that the user is not aware which of the realms in a domain successfully authenticated him).
We haven't discussed it throughly, but I don't think there is a way that we can abstract away the owned/visible separation without also ending up with resources that are shared across domains, which is also not desirable.

----

What is still left to discuss is to more accurately explain the behavior for resources that have been modified by different realms from the same domain, after the realms have been separated between different domains (technically both can claim ownership of the resource, but the resource cannot? be accessible by different domains).
**In general, we need to make it clearer how are the resources divided when a domain is divided.**

**Comment by albertzaharovits** [comment]
Today @ywangd and I had a productive chat about the changes to the `Authentication` object to reflect the domain assignment of the role.

The TLDR is that the `Authentication` object will hold the complete realm refs of the domains, rather than the domain name.

Rationale:

The user (subject) and the resource (object) both have `Authentication`s attached. These `Authentication`s are the only pieces of information that are compared in order to authorize the access. Importantly, they are created at different times (eg when the scroll/API Key is created vs when it is used, eg possibly by an ML scheduled job) and places (different coordinating and data nodes can technically have different realms/domain configuration). At a high level, if we only refer to a domain by its name, we make the authorization check less predictable, as the check will have to inspect the current domain configuration on the current node, which can be completely different from when the resources were created or when/where the user authenticated. This is analogous to the current authentication idiosyncracy with role names being stored rather than role permissions.

In order to simplify this process the user authentication will contain the realm refs (name/type tuple), avoiding the domain name altogether. Such an `Authentication` (containing the realm refs for the domain) is then copied over to the resources that the user creates. When the user accesses an existing resource, the realm refs stored against that resource's authentication are used in the authorization decision, again avoiding to refer to any domain names.
The domain name is hence completely fluid, it can change without any impact to security.
Realm names and types however must continue to be "set in stone". Realm rename/retype as well as breaking up existing domains (by reassigning realms) is going to make resources inaccessible.

NB: In practice, currently, the authorization decision on accessing resources is only going to inspect the domain realm refs of the user's authentication (and ignoring the resource's). In the future, we might devise resources that "stick to a domain" for which authorizaiton will look at the domain assignment on the resource rather than on the user.

**Comment by albertzaharovits** [comment]
This work has been incorporated in various other PRs (that I've lost track off).
